ZelBack/src/services/appsService.js

```diff
@@ -14203,49 +14284,92 @@ async function checkMyAppsAvailability() {
-        log.info(`checkMyAppsAvailability - Testing port ${testingPort} failed to create on UPNP mappings. Possible already assigned?`);
-        failedPort = null;
-        if (originalPortFailed && testingPort > originalPortFailed) {
-          setPortToTest = originalPortFailed - 1;
-        } else if (originalPortFailed) {
-          setPortToTest = null;
-        }
-        throw new Error('Failed to create map UPNP port');
+        log.info(
+          `checkMyAppsAvailability - Testing port ${testingPort} failed to create on UPnP mappings`,
+        );
+
+        setNextPort();
+
+        await handleTestShutdown(testingPort, {
+          skipFirewall: !firewallActive,
+          skipUpnp: true,
+          skipHttpServer: true,
+        });
+        // set a longer delay here if we are DOS (from UPnP or any other reason).
+        // We don't want to sit there hammering the router every 15 seconds
+        // if mappings are failing, especially if there are 8 nodes going at it
+        const upnpDelay = dosMessage ? 2 * 60 * 1000 : 15 * 60 * 1000;
+        await serviceHelper.delay(upnpDelay);
+        checkMyAppsAvailability();
+        return;
       }
       lastUPNPMapFailed = false;
     }
+
     await serviceHelper.delay(5 * 1000);
-    testingAppserver.listen(testingPort).on('error', (err) => {
-      throw err.message;
-    }).on('uncaughtException', (err) => {
-      throw err.message;
-    });
+
+    // these event listeners need to be removed on every iteration to prevent a
+    // memory leak
+    testingAppserver
+      .listen(testingPort)
+      .on('error', (err) => {
+        throw err.message;
+      })
+      .on('uncaughtException', (err) => {
+        throw err.message;
+      });
+
     await serviceHelper.delay(10 * 1000);
-    // eslint-disable-next-line no-await-in-loop
+
     let askingIP = await fluxNetworkHelper.getRandomConnection();
+
     if (!askingIP) {
+      await handleTestShutdown(testingPort, {
+        skipFirewall: !firewallActive,
+        skipUpnp: !isUpnp,
+      });
+      // changed this to 4 minutes. If we can't get a random connection, we have
+      // other problems. (i.e. it's an error state, or node just started)
+      await serviceHelper.delay(4 * 60 * 1000);
       checkMyAppsAvailability();
       return;
     }
+
     let askingIpPort = config.server.apiport;
-    if (askingIP.includes(':')) { // has port specification
-      // it has port specification
+    if (askingIP.includes(':')) {
       const splittedIP = askingIP.split(':');
       askingIP = splittedIP[0];
       askingIpPort = splittedIP[1];
     }
+
     if (myIP === askingIP) {
+      await handleTestShutdown(testingPort, {
+        skipFirewall: !firewallActive,
+        skipUpnp: !isUpnp,
+      });
+      // safer just to wait here, no harm in waiiting 15 seconds in unlikely event
+      // that we pull ourselves from the list (or we should just remove it first)
+      await serviceHelper.delay(15 * 60 * 1000);
```

Comment:
this is awaiting 15 minutes, not 15 seconds

Reply:
lol. Yes. Will check it out tomorrow

Reply:
Updated to 15 seconds